# Analisis de datos (AcuaVentura)

### Instalaciones y importaciones necesarias

In [1420]:
pip install xlrd

Note: you may need to restart the kernel to use updated packages.


In [1421]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
df = pd.read_excel("GSAF5.xls", engine="xlrd")

In [1422]:
pd.set_option('display.max_columns', None)
#Esto es para ver todas las columnas cuando hacemos un head, un describe... etc.

### Inspeccion!!

In [1423]:
df.head()
#Parece que hay muchos valores muy sucios dentro de las columnas y algunas columnas parece que tienen muchos nulos.

Date    Year        Type    Country                State  \
0  2025-04-19 00:00:00  2025.0  Unprovoked   Maldives  Kulhudhuffushi City   
1  2025-04-12 00:00:00  2025.0  Unprovoked        USA              Florida   
2  2025-03-26 00:00:00  2025.0  Unprovoked  Australia                   WA   
3  2025-03-10 00:00:00  2025.0  Unprovoked  Australia                   WA   
4  2025-03-07 00:00:00  2025.0  Unprovoked  Australia                  NSW   

                                            Location     Activity  \
0                                   Haa Dhaalu atoll   snorkeling   
1  Everglades National Park Flamingo Lodge Highwa...  Undisclosed   
2          Sandtrax Port Beach North Fremantle Perth     Swimming   
3                                Duke of Orleans Bay      Surfing   
4                 Gunyah beach Bundeena Port Hacking     Swimming   

                   Name Sex  Age                                 Injury  \
0          Unknown Male   M   30             Injuries to right shoulder   
1          Unknown Male   M    ?         Injuries to right hand and arm   
2          Unknown Male   M  30+                    Bumping to the body   
3  Steven Jeffrey Payne   M   37             Missing body not recovered   
4        Mangyong Zhang  F    56  Serious lacerations above right knee.   

  Fatal Y/N        Time           Species   \
0         N  Not stated         Not stated   
1         N     1500hrs         Not stated   
2         N     1430hrs   1.5m Tiger shark   
3         Y    1210 hrs  Great White Shark   
4         N     1340hrs         Bull shark   

                                              Source  pdf href formula href  \
0   Todd Smith: The edition: https://en.sun.mv/96031  NaN          NaN  NaN   
1  Kevin McMurray Trackingsharks.com: Florida New...  NaN          NaN  NaN   
2  Kevin McMurray Trackingsharks.com: www.surfer....  NaN          NaN  NaN   
3                                          Bob Myatt  NaN          NaN  NaN   
4                                          Bob Myatt  NaN          NaN  NaN   

  Case Number Case Number.1  original order Unnamed: 21 Unnamed: 22  
0         NaN           NaN             NaN         NaN         NaN  
1         NaN           NaN             NaN         NaN         NaN  
2         NaN           NaN             NaN         NaN         NaN  
3         NaN           NaN             NaN         NaN         NaN  
4         NaN           NaN             NaN         NaN         NaN

In [1424]:
df.shape

(7012, 23)

In [1425]:
df.columns
#Hay que limpiar los nombres de las columnas.

Index(['Date', 'Year', 'Type', 'Country', 'State', 'Location', 'Activity',
       'Name', 'Sex', 'Age', 'Injury', 'Fatal Y/N', 'Time', 'Species ',
       'Source', 'pdf', 'href formula', 'href', 'Case Number', 'Case Number.1',
       'original order', 'Unnamed: 21', 'Unnamed: 22'],
      dtype='object')

In [1426]:
df.dtypes
#Hay que cambiar los tipos de dato, hemos decidido dejarlo para más tarde, cuando hayamos limpiado bien las columnas.

Date               object
Year              float64
Type               object
Country            object
State              object
Location           object
Activity           object
Name               object
Sex                object
Age                object
Injury             object
Fatal Y/N          object
Time               object
Species            object
Source             object
pdf                object
href formula       object
href               object
Case Number        object
Case Number.1      object
original order    float64
Unnamed: 21        object
Unnamed: 22        object
dtype: object

In [1427]:
df.describe().T

count         mean          std  min     25%     50%     75%  \
Year            7010.0  1935.271327   271.735460  0.0  1947.0  1986.0  2009.0   
original order  6799.0  3401.152081  1963.076319  2.0  1701.5  3401.0  5100.5   

                   max  
Year            2026.0  
original order  6802.0

In [1428]:
df.describe(include= "object")

Date        Type Country    State                          Location  \
count   7012        6994    6962     6527                              6446   
unique  6057          12     242      926                              4570   
top     1957  Unprovoked     USA  Florida  New Smyrna Beach, Volusia County   
freq       9        5170    2558     1186                               191   

       Activity  Name   Sex   Age Injury Fatal Y/N       Time     Species   \
count      6427  6793  6433  4018   6977      6451       3486         3881   
unique     1604  5742    10   249   4139        12        444         1712   
top     Surfing  male     M    16  FATAL         N  Afternoon  White shark   
freq       1133   679  5624    92    863      4883        215          194   

                                 Source                        pdf  \
count                              6993                       6799   
unique                             5356                       6789   
top     K. McMurray, TrackingSharks.com  1907.10.16.R-HongKong.pdf   
freq                                131                          2   

                                             href formula  \
count                                                6794   
unique                                               6784   
top     http://sharkattackfile.net/spreadsheets/pdf_di...   
freq                                                    2   

                                                     href Case Number  \
count                                                6796        6798   
unique                                               6776        6777   
top     http://sharkattackfile.net/spreadsheets/pdf_di...  2021.07.23   
freq                                                    4           2   

       Case Number.1   Unnamed: 21 Unnamed: 22  
count           6797             1           2  
unique          6775             1           2  
top     2012.09.02.b  stopped here      Teramo  
freq               2             1           1

### Limpieza de columnas

In [1429]:
#Empezamos con columnas, en str, minusculas y reemplazando espacio por _.
df.columns = (df.columns.str.strip().str.lower().str.replace(" ", "_"))

In [1430]:
#Hacemos que todas sean faciles de nombrar.
df.columns = df.columns.str.replace("fatal_y/n", "fatal")

In [1431]:
#Eliminamos las que no nos interesan, otras las eliminaremos más tarde para poder eliminar duplicados.
df = df.drop(['location','state','source', 'pdf', 'href_formula', 'href', 'case_number', 'case_number.1', 'original_order', 'unnamed:_21', 'unnamed:_22'], axis=1)

In [1432]:
#Comprobamos si esta todo bien.
df.columns

Index(['date', 'year', 'type', 'country', 'activity', 'name', 'sex', 'age',
       'injury', 'fatal', 'time', 'species'],
      dtype='object')

### Limpiamos las filas de cada columna

In [1433]:
df['sex'].unique()

array(['M', 'F ', 'F', 'M ', nan, ' M', 'm', 'lli', 'M x 2', 'N', '.'],
      dtype=object)

In [1434]:
#limpiamos el sexo de cada victima
df['sex'] = df['sex'].map({
    'F ': 'F',
    'F': 'F',
    '.': "F",
    'M': 'M',
    'M ': 'M',
    ' M': 'M',
    'm': 'M',
    'lli': 'M',
    'M x 2': 'M',
    'N': 'M'
    })
df['sex'].unique()

array(['M', 'F', nan], dtype=object)

In [1435]:
#hay muchos paises vamos a limpiarlos
df['country'].unique()

array(['Maldives', 'USA', 'Australia', 'Philippines', 'Bahamas',
       'Turks and Caicos', 'Mozambique', 'New Caledonia', 'Egypt',
       'Thailand', 'New Zealand', 'Hawaii', 'Honduras', 'Indonesia',
       'Morocco', 'Jamaica', 'Belize', 'Maldive Islands',
       'French Polynesia', 'Tobago', 'AUSTRALIA', 'INDIA', 'TRINIDAD',
       'BAHAMAS', 'SOUTH AFRICA', 'MEXICO', 'NEW ZEALAND', 'EGYPT',
       'Mexico', 'BELIZE', 'PHILIPPINES', 'Coral Sea', 'SPAIN',
       'PORTUGAL', 'SAMOA', 'COLOMBIA', 'ECUADOR', 'FRENCH POLYNESIA',
       'NEW CALEDONIA', 'TURKS and CaICOS', 'CUBA', 'BRAZIL',
       'SEYCHELLES', 'ARGENTINA', 'FIJI', 'MeXICO', 'South Africa',
       'ENGLAND', 'JAPAN', 'INDONESIA', 'JAMAICA', 'MALDIVES', 'THAILAND',
       'COLUMBIA', 'COSTA RICA', 'British Overseas Territory', 'CANADA',
       'JORDAN', 'ST KITTS / NEVIS', 'ST MARTIN', 'PAPUA NEW GUINEA',
       'REUNION ISLAND', 'ISRAEL', 'CHINA', 'IRELAND', 'ITALY',
       'MALAYSIA', 'LIBYA', nan, 'MAURITIUS', 'SOLOMON 

In [1436]:
#Primero las igualamos.
df['country'] = df['country'].astype(str).str.strip().str.lower()
#Ahora empezamos el mapeo.
mapeo_paises = {
    # --- UNITED STATES ---
    'usa': 'united states',
    'hawaii': 'united states',
    'johnston island': 'united states',
    'guam': 'united states',
    'puerto rico': 'united states',
    'northern mariana islands': 'united states',
    'johnston atoll': 'united states',
    'american samoa': 'united states',

    # --- UNITED KINGDOM ---
    'england': 'united kingdom',
    'scotland': 'united kingdom',
    'british isles': 'united kingdom',
    'saint helena': 'united kingdom',
    'cayman islands': 'united kingdom',
    'turks and caicos islands': 'united kingdom',
    'british indian ocean territory': 'united kingdom',
    'british virgin islands': 'united kingdom',
    'bermuda': 'united kingdom',
    'british overseas territory': 'united kingdom',
    'falkland islands': 'united kingdom',

    # --- FRANCE ---
    'reunion': 'france',
    'reunion island': 'france',
    'new caledonia': 'france',
    'french polynesia': 'france',
    'mayotte': 'france',
    'martinique': 'france',
    'saint martin': 'france',

    # --- NETHERLANDS ---
    'aruba': 'netherlands',
    'curacao': 'netherlands',
    'netherlands antilles': 'netherlands',

    # --- PORTUGAL ---
    'azores': 'portugal',

    # --- INDIA ---
    'andaman islands': 'india',
    'andaman / nicobar islandas': 'india',
    'indian ocean?': 'india',
    'west indies': None, 

    # --- SRI LANKA ---
    'ceylon': 'sri lanka',
    'ceylon (sri lanka)': 'sri lanka',

    # --- SAMOA ---
    'western samoa': 'samoa',

    # --- JAPAN ---
    'okinawa': 'japan',

    # --- SOUTH KOREA ---
    'korea': 'south korea',

    # --- PALESTINE ---
    'palestinian territories': 'palestine',
    'palestine': 'palestine',

    # --- COLOMBIA ---
    'columbia': 'colombia',

    # --- TRINIDAD & TOBAGO ---
    'trinidad': 'trinidad and tobago',
    'tobago': 'trinidad and tobago',
    'trinidad & tobago': 'trinidad and tobago',

    # --- HONDURAS ---
    'roatan': 'honduras',

    # --- CABO VERDE ---
    'cape verde': 'cabo verde',

    # --- MYANMAR ---
    'burma': 'myanmar',

    # --- BAHRAIN ---
    'bahrein': 'bahrain',

    # --- MICRONESIA ---
    'federated states of micronesia': 'micronesia',

    # --- PAPUA NEW GUINEA ---
    'new britain': 'papua new guinea',
    'admiralty islands': 'papua new guinea',
    'british new guinea': 'papua new guinea',
    'new guinea': 'papua new guinea',

    # --- GREECE ---
    'crete': 'greece',

    # --- DENMARK ---
    'greenland': 'denmark',

    # --- CHINA ---
    'hong kong': 'china',

    # --- OCEAN / REGIONES MARÍTIMAS AMBIGUAS --- (todo esto se convierte en None)
    'pacific ocean': None,
    'atlantic ocean': None,
    'caribbean sea': None,
    'coral sea': None,
    'tasman sea': None,
    'north atlantic ocean': None,
    'north sea': None,
    'south china sea': None,
    'mid-pacifc ocean': None,
    'mid atlantic ocean': None,
    'mediterranean sea': None,
    'gulf of aden': None,
    'bay of bengal': None,
    'persian gulf': None,
    'red sea': None,
    'red sea?': None,
    'indian ocean': None,
    'indian ocean?': None,
    'south atlantic ocean': None,
    'south pacific ocean': None,
    'north pacific ocean': None,
    'central pacific': None,
    'southwest pacific ocean': None,
    'northern arabian sea': None,
    'red sea / indian ocean': None,
    'sudan?': 'sudan',

    # --- REGIONES GEOGRÁFICAS INDEFINIDAS ---
    'coast of africa': None,
    'africa': None,
    'the balkans': None,
    'ocean': None,
    'between portugal & india': None,
    'british west indies': None,
    'asia?': None,
    'equatorial guinea / cameroon': None,
    'egypt / israel': None,
    'iran / iraq': None,
    'italy / croatia': None,
    'solomon islands / vanuatu': None,

    # --- ISLAS O TERRITORIOS MENORES (claros) ---
    'cook islands': 'new zealand',
    'cook island': None,  
    'maldive islands': 'maldives',
    'turks and caicos': 'turks and caicos islands',
    'turks & caicos': 'turks and caicos islands',
    'united arab emirates (uae)': 'united arab emirates',
    'st martin': 'saint martin',
    'st. martin': 'saint martin',
    'st. maartin': 'saint martin',
    'grand cayman': 'cayman islands',
    'nevis': 'saint kitts and nevis',
    'st kitts / nevis': 'saint kitts and nevis',
    'diego garcia': 'british indian ocean territory',
    'java': 'indonesia',
    'san domingo': 'dominican republic',

    # --- NULOS / INDETERMINADOS ---
    'nan': None,
    None: None
}

#Aplicamos el mapeo.
df['country'] = df['country'].replace(mapeo_paises)
df['country'].unique()

array(['maldives', 'united states', 'australia', 'philippines', 'bahamas',
       'turks and caicos islands', 'mozambique', 'france', 'egypt',
       'thailand', 'new zealand', 'honduras', 'indonesia', 'morocco',
       'jamaica', 'belize', 'trinidad and tobago', 'india',
       'south africa', 'mexico', None, 'spain', 'portugal', 'samoa',
       'colombia', 'ecuador', 'cuba', 'brazil', 'seychelles', 'argentina',
       'fiji', 'united kingdom', 'japan', 'costa rica', 'canada',
       'jordan', 'saint kitts and nevis', 'saint martin',
       'papua new guinea', 'israel', 'china', 'ireland', 'italy',
       'malaysia', 'libya', 'mauritius', 'solomon islands',
       'st helena, british overseas territory', 'comoros',
       'united arab emirates', 'cabo verde', 'dominican republic',
       'netherlands', 'greece', 'kiribati',
       'british indian ocean territory', 'taiwan', 'palestine', 'nigeria',
       'tonga', 'croatia', 'saudi arabia', 'chile', 'antigua', 'kenya',
       'russia',

In [1437]:
#Filtramos los paises de nuestra empresa.
paises_validos = [
    'spain', 'portugal', 'greece', 'croatia',
    'thailand', 'indonesia', 'philippines', 'japan', 'mexico', 'peru', 'cuba'
]
df = df[df['country'].isin(paises_validos)]
df['country'].unique()

array(['philippines', 'thailand', 'indonesia', 'mexico', 'spain',
       'portugal', 'cuba', 'japan', 'greece', 'croatia', 'peru'],
      dtype=object)

In [1438]:
#Limpiamos activity, todas sean str, sin espacios y en minusculas.
df['activity'] = df['activity'].astype(str).str.strip().str.lower()
df['activity'].nunique()

160

In [1439]:
#Filtramos las que contienen ciertas palabras para poder hacer una nueva clasificación.
df['activity_clean'] = df['activity']
df.loc[df['activity'].str.contains('swim|bathing|wading|floating|playing|jump|train', na=False), 'activity_clean'] = 'swimming'
df.loc[df['activity'].str.contains('diving|snorkel', na=False), 'activity_clean'] = 'diving'
df.loc[df['activity'].str.contains('surf|kite|board|kayak|cano|rowing|padd', na=False), 'activity_clean'] = 'surf'
df.loc[df['activity'].str.contains('fish|boat', na=False), 'activity_clean'] = 'fishing'
df.loc[df['activity'].str.contains('wreck|disaster', na=False), 'activity_clean'] = 'wreck'
categorias = ['swimming', 'diving', 'surf', 'fishing', 'wreck']
df.loc[~df['activity_clean'].isin(categorias), 'activity_clean'] = 'other'
df['activity'] = df['activity_clean']
df = df.drop(['activity_clean'], axis=1)
df['activity'].nunique()

6

In [1440]:
#Limpiamos injury, todas sean str, sin espacios y en minusculas.
df['injury'] = df['injury'].astype(str).str.strip().str.lower()
#Metemos los valores en categorias nuevas como en el anterior pero de una forma distinta.
def categorize_injury(injury):
    if pd.isna(injury):
        return 'Unknown'

    injury = injury.lower()

    if 'fatal' in injury or 'died' in injury or 'death' in injury:
        return 'Fatal'
    elif 'serious' in injury or 'severed' in injury or 'critical' in injury:
        return 'Serious injury'
    elif 'no injury' in injury or 'uninjured' in injury or 'not injured' in injury or 'no' in injury:
        return 'No injury'
    elif 'minor' in injury or 'superficial' in injury or 'lower' in injury:
        return 'Minor injury'
    elif 'leg' in injury or 'foot' in injury or 'thigh' in injury or 'ankle' in injury or 'calf' in injury:
        return 'Injury to the lower trunk'
    elif 'arm' in injury or 'hand' in injury:
        return 'Injury to the upper trunk'
    elif 'lacerations' in injury or 'lacerated' in injury or 'laceration' in injury:
        return 'Laceration'
    elif 'injury' in injury or 'bitten' in injury:
        return 'Unknown type of injury'
    elif 'unknown' in injury or 'n/a' in injury:
        return 'Unknown'
    else:
        return 'Other'

df['injury'] = df['injury'].apply(categorize_injury)

In [1441]:
#Limpiamos fatal, todas sean str, sin espacios y en mayusculas.
df['fatal'] = df['fatal'].astype(str).str.strip().str.upper()
#Tambien usamos map para cambiar las Y a YES y N a NO
df['fatal'] = df['fatal'].map({'Y': 'Yes', 'N': 'No',})
df['fatal'].unique()

array(['Yes', 'No', nan], dtype=object)

In [1442]:
#Nos aseguramos de que todos los valores sean strings limpios.
df['time'] = df['time'].astype(str).str.strip().str.upper()
df['hour'] = df['time'].str.extract(r'(\d{1,2})')
#ver distribución de horas
df['hour'].value_counts().sort_index() 

hour
02     1
03     1
04     1
05     1
07     6
08     3
09     2
10    11
11     8
12     6
13     6
14     9
15     9
16     5
17     7
18     5
19     4
23     2
Name: count, dtype: int64

In [1443]:
# Primero aseguramos que hour es string 
df['hour'] = df['hour'].astype(str).str.zfill(2)
# Luego creamos una nueva columna en formato HH:00
df['hour_formatted'] = df['hour'] + ':00'
df['hour_formatted'].value_counts().sort_index()
df['hour'] = df['hour_formatted']
df = df.drop(['hour_formatted','time'], axis=1)
df['hour'].value_counts().sort_index() 

hour
02:00       1
03:00       1
04:00       1
05:00       1
07:00       6
08:00       3
09:00       2
10:00      11
11:00       8
12:00       6
13:00       6
14:00       9
15:00       9
16:00       5
17:00       7
18:00       5
19:00       4
23:00       2
nan:00    332
Name: count, dtype: int64

In [1444]:
#Limpiamos fechas pasandolas a date time, extraemos month para crear una nueva columna y llenamos los años con el resto de informacion de date.
def limpiar_fechas(df):
    df['date'] = pd.to_datetime(df['date'], errors='coerce')
    df['month'] = df['date'].dt.month
    df['year_from_date'] = df['date'].dt.year
    df['year'] = df['year'].fillna(df['year_from_date'])
    df = df.drop(columns=['date', 'year_from_date'])

    return df
df=limpiar_fechas(df)

C:\Users\rocio\AppData\Local\Temp\ipykernel_13812\1983724394.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['date'] = pd.to_datetime(df['date'], errors='coerce')


In [1445]:
#Sustituimos los datos sucios con sus nombres respectivos.

type_r = {
    ' Provoked': 'Provoked',
    'Questionable': 'Unknown',
    '?': 'Unknown',
    'Unconfirmed': 'Unknown',
    'Unverified': 'Unknown',
    'Invalid': 'Unknown',
    'Under investigation': 'Unknown',
    'Boat': 'Watercraft',
    'Sea Disaster': 'Watercraft'  
}
df['type'] = df['type'].replace(type_r)
df['type'] = df['type'].str.lower().str.strip()
print(df['type'].unique())

['unknown' 'unprovoked' 'provoked' 'watercraft' nan]


In [1446]:
#Limpiamos las especies de tiburón.
df['species'] = df['species'].str.lower().str.strip()
species_r = {
    'tiger shark': 'Tiger shark',
    "1.5m tiger shark": 'Tiger shark',
    "12' tiger shark": 'Tiger shark',
    'great white shark': 'Great white shark',
    'bull shark': 'Bull shark'
}
df['species'] = df['species'].replace(species_r)
new_species = []
for value in df['species']:
    if value == 'Tiger shark':
        new_species.append('Tiger shark')
    elif value == 'Great white shark':
        new_species.append('Great white shark')
    elif value == 'Bull shark':
        new_species.append('Bull shark')
    else:
        new_species.append('Other')
df['species'] = new_species
df['species'].unique()

array(['Other', 'Bull shark', 'Tiger shark'], dtype=object)

In [1447]:
#Cambiamos limpiamos los años.
df['age'] = df['age'].astype(str).str.strip()
df['age'] = df['age'].str.replace(r'[^\d]', '', regex=True)
df['age'].unique()

array(['29', '57', '68', '22', '66', '52', '26', '59', '', '35', '17',
       '10', '53', '60', '8', '56', '14', '55', '12', '9', '75', '23',
       '32', '54', '13', '42', '11', '82', '33', '25', '48', '40', '24',
       '38', '37', '36', '16', '30', '43', '31', '34', '2826', '46', '18',
       '49', '15', '39', '21', '7', '19', '51', '28', '47', '69', '41',
       '20', '45', '58', '1318', '50', '78', '3623', '27'], dtype=object)

In [1448]:
#Como solo queremos dos digitos como mucho le aplicamos una funcion.
#Esta se queda solo con los dos primeros digitos.
def clean_age(age):
    if isinstance(age, str) and len(age) > 2 and age.isdigit():
        return age[:2]
    else:
        return age
df['age'] = df['age'].apply(clean_age)
df['age'] = df['age'].replace(r'^\s*$', np.nan, regex=True)
df['age'].unique()

array(['29', '57', '68', '22', '66', '52', '26', '59', nan, '35', '17',
       '10', '53', '60', '8', '56', '14', '55', '12', '9', '75', '23',
       '32', '54', '13', '42', '11', '82', '33', '25', '48', '40', '24',
       '38', '37', '36', '16', '30', '43', '31', '34', '28', '46', '18',
       '49', '15', '39', '21', '7', '19', '51', '47', '69', '41', '20',
       '45', '58', '50', '78', '27'], dtype=object)

In [1449]:
#Ya las tenemos todas limpias, falta ordenarlas.
columnas_ordenadas = ['year','month','hour','country','type','activity','name','sex','age','injury','fatal','species']
df = df[columnas_ordenadas]

In [1450]:
#Porfin cambiamos los tipos de dato.
df['year'] = df['year'].astype('Int64')
df['month'] = df['month'].astype('Int64')
df['age'] = pd.to_numeric(df['age'], errors='coerce')
df['age'] = df['age'].astype('Int64')

### Duplicados

In [1451]:
#No hay duplicados.
df.duplicated().sum()

0

In [1452]:
#Podemos ya eliminar name.
df = df.drop(['name'], axis=1)

### Valores nulos

In [1453]:
df.shape

(419, 11)

In [1454]:
#Parece que hay unos cuantos vamos a ver como solucionarlos.
df.isna().sum()

year          0
month        96
hour          0
country       0
type          2
activity      0
sex          62
age         250
injury        0
fatal        57
species       0
dtype: int64

In [1455]:
#Eliminamos columnas completamente vacías por si acaso.
df.dropna(axis=1, how='all', inplace=True)

In [1456]:
#Eliminamos filas completamente vacías por si acaso.
df.dropna(axis=0, how='all', inplace=True)

In [1457]:
#El mes hemos decidido llenarlo con la moda.
df['month'] = df['month'].fillna(df['month'].mode()[0])
#El resto con unknown ya que no nos gustaria que nos diera información que no es.
df['country'] = df['country'].fillna('unknown')
df['type'] = df['type'].fillna('unknown')
df['sex'] = df['sex'].fillna('unknown')
df['age'] = df['age'].fillna(999)
df['fatal'] = df['fatal'].fillna('unknown')

### Comenzamos a sacar insights

Nuestra intencion es hacer que las actividades de nuestra empresa sean más seguras, por lo tanto vamos a intentar sacar información para saber donde se necesita reforzar o hacer cosas distintas para disminuir los ataques de tiburones.

In [1458]:
# Vamos a descubrir primero en que años se colocan nuestros datos de ataques.
años_validos = df[df['year'] > 1000]

año_menor = años_validos['year'].min()
año_mayor = años_validos['year'].max()

print("Año más antiguo:", año_menor)
print("Año más reciente:", año_mayor)

Año más antiguo: 1500
Año más reciente: 2025


In [1459]:
#Parece un rango muy extenso, voy a ver cual fue el año con mas ataques de cada pais.
pivot = df[df['year'] > 1000].pivot_table(
    index='country',
    columns='year',
    values='activity',  
    aggfunc='count',
    fill_value=0
)
pivot['year_mas_ataques'] = pivot.idxmax(axis=1)
resultado = pivot[['year_mas_ataques']].reset_index()
print(resultado)

year      country  year_mas_ataques
0         croatia              1907
1            cuba              1957
2          greece              1951
3       indonesia              1906
4           japan              1992
5          mexico              2023
6            peru              1849
7     philippines              1944
8        portugal              1872
9           spain              1862
10       thailand              2018


· Ciertos países (como México 2023 o Tailandia 2018) muestran picos recientes en ataques:

→ Reforzar protocolos de seguridad en esas ubicaciones en temporadas similares.

In [1460]:
#¿que rangos de edad sufren más ataques?
df_clean = df[df['age'].between(5, 100)] 
#Vamos a crear primero rangos de edad para averiguarlo.
df_clean['age_group'] = pd.cut(df_clean['age'], bins=[0, 20, 35, 50, 65, 100], labels=["0-20", "21-35", "36-50", "51-65", "66+"])
df['age'] = df_clean['age_group']
df_clean.groupby('age_group').size().sort_values(ascending=False)

C:\Users\rocio\AppData\Local\Temp\ipykernel_13812\3385326374.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['age_group'] = pd.cut(df_clean['age'], bins=[0, 20, 35, 50, 65, 100], labels=["0-20", "21-35", "36-50", "51-65", "66+"])
C:\Users\rocio\AppData\Local\Temp\ipykernel_13812\3385326374.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_clean.groupby('age_group').size().sort_values(ascending=False)


age_group
21-35    60
0-20     54
36-50    32
51-65    16
66+       7
dtype: int64

· El grupo 21-35 años es el más afectado por ataques de tiburón (60 casos), seguido por 0-20 años (54 casos).
Estos dos grupos representan más del 70% de los ataques.

→ Priorizar formación preventiva y protocolos de seguridad para jóvenes y adultos jóvenes.

→ Tal vez limitar rangos de edad a ciertas actividades

In [1461]:
comparativa = df.groupby(['age', 'fatal']).size().reset_index(name='count')
pivot_edad_fatal = comparativa.pivot(index='age', columns='fatal', values='count').fillna(0)
pivot_edad_fatal['fatality_rate_%'] = (pivot_edad_fatal['Yes'] / pivot_edad_fatal.sum(axis=1)) * 100
pivot_edad_fatal = pivot_edad_fatal.round(1)
pivot_edad_fatal

C:\Users\rocio\AppData\Local\Temp\ipykernel_13812\3364594292.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  comparativa = df.groupby(['age', 'fatal']).size().reset_index(name='count')


fatal  No  Yes  unknown  fatality_rate_%
age                                     
0-20   24   22        8             40.7
21-35  26   30        4             50.0
36-50  19    9        4             28.1
51-65   8    7        1             43.8
66+     3    3        1             42.9

21-35 años tienen la mayor tasa de mortalidad (50%). Aunque el número de ataques es alto, el riesgo de muerte es notable,
los grupos 51-65 años (43.8%) y 66+ años (42.9%) tienen tasas de fatalidad elevadas a pesar de tener menos ataques. Son altas excepto en 36-50.

→ Refuerza la idea de limitar los rangos de edad de ciertas actividades, icluso incluyendo in limite de edad.

→ Si fuera asumir muchas perdidas se tendrian que priorizar medidas de seguridad en jovenes y en adultos mayores, se podrian añadir guias adicionales y monitoreo mas constante.

In [1462]:
#Actividades más peligrosas:
df.groupby('activity').size().sort_values(ascending=False)

activity
swimming    144
other        98
fishing      86
diving       40
surf         38
wreck        13
dtype: int64

Swimming (144 incidentes) es la actividad con más ataques, lo que indica que el riesgo es elevado en actividades acuáticas sin protección. 

→ Evaluar crear protocolos de seguridad al nadar.

In [1463]:
#En que rango de edad existe mas riesgo en cada actividad.
df.groupby(['activity','age']).size().sort_values(ascending=False).head()

C:\Users\rocio\AppData\Local\Temp\ipykernel_13812\1628908118.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(['activity','age']).size().sort_values(ascending=False).head()


activity  age  
swimming  0-20     33
          21-35    26
diving    21-35    10
swimming  36-50    10
fishing   36-50     9
dtype: int64

In [1464]:
#Filtrando los paises en los que tenemos nuestras actividades, ¿en que actividades tenemos que incidir mas en la seguridad segun el pais?
resumen = df.groupby(['country', 'activity']).size().reset_index(name='incident_count')
resumen_ordenado = resumen.sort_values(by=['country', 'incident_count'], ascending=[True, False])
mas_peligrosa_por_pais = resumen_ordenado.groupby('country').first().reset_index()
print(mas_peligrosa_por_pais)

        country  activity  incident_count
0       croatia  swimming              15
1          cuba   fishing              16
2        greece  swimming              11
3     indonesia      surf               8
4         japan     other              16
5        mexico  swimming              44
6          peru     other               1
7   philippines     other              21
8      portugal   fishing               7
9         spain  swimming              25
10     thailand  swimming               7


5 de 11 paises en los que tenemos nuestras actividades también tienen un número considerable de incidentes en swimming, lo que sugiere que esta actividad es una de las más peligrosas en diversas ubicaciones.

In [1465]:
#Vamos a ver en general que pais de nuestras sedes es más peligroso.
df.groupby('country').size().sort_values(ascending=False).head(10)

country
mexico         106
philippines     65
cuba            49
spain           46
japan           42
croatia         35
greece          26
indonesia       26
thailand        13
portugal         9
dtype: int64

In [1466]:
#¿Existe algun riesgo dependiendo de la hora?
df.groupby('hour').size().sort_values(ascending=False).head(10)

hour
nan:00    332
10:00      11
15:00       9
14:00       9
11:00       8
17:00       7
13:00       6
12:00       6
07:00       6
16:00       5
dtype: int64

Parece que no tenemos los datos suficientes para asegurarnos ya que hay muchos que no conocemos
pero podemos empezar a aplicar mas actividades en la tarde, a partir de las 6.

In [1467]:
#¿Hay diferencias de genero?
df.groupby('sex').size().sort_values(ascending=False)

sex
M          312
unknown     62
F           45
dtype: int64

Parece que estos datos nos llaman la atencion, vamos a investigar un poco mas.

In [1468]:
pivot_fatalidad_1 = df.pivot_table(index='activity', columns='sex', aggfunc='size', fill_value=0)
pivot_fatalidad_1

sex        F    M  unknown
activity                  
diving     2   36        2
fishing    2   65       19
other      6   63       29
surf       3   34        1
swimming  32  106        6
wreck      0    8        5

Podriamos enfocar medidas de seguridad en hombres, ya que son el grupo más afectado en actividades como swimming y fishing.
Potenciar más control y supervisión en actividades con alta participación masculina, especialmente en swimming.

In [1469]:
#¿Que grupo muere más cuando se trata de cada actividad?¿Hay algun grupo con mas tendencia?
pivot_fatalidad = df.pivot_table(index='activity', columns=['sex', 'fatal'], aggfunc='size', fill_value=0)
pivot_fatalidad

sex        F               M             unknown            
fatal     No Yes unknown  No Yes unknown      No Yes unknown
activity                                                    
diving     0   2       0  13  17       6       1   0       1
fishing    0   2       0  38  23       4      11   6       2
other      2   4       0  17  33      13       6   8      15
surf       2   1       0  17  14       3       0   1       0
swimming  14  15       3  45  53       8       1   4       1
wreck      0   0       0   3   4       1       0   5       0

En la mayoría de las actividades, hombres tienen más muertes, destacando:

Fishing (23 muertes en hombres).

Swimming (53 muertes en hombres).

Other (33 muertes en hombres).

→Podriamos generar formación preventiva más centrada en hombres para dar una mayor concienciación sobre estos datos.

In [1470]:
#¿Hay alguna especie más involucrada?
df.groupby('species').size().sort_values(ascending=False)

species
Other          409
Bull shark       5
Tiger shark      5
dtype: int64

No tenemos los datos suficientes

In [1471]:
#¿Hay risgo en la estacionalidad?
df.groupby('month').size().sort_values(ascending=False)

month
7     147
1      44
8      43
9      31
5      25
11     25
10     22
6      21
4      20
3      17
12     16
2       8
dtype: int64

Parece haber un pico alto de ataques en el verano, con Julio, Agosto y Septiembre siendo los meses más peligrosos.

Los meses de invierno (Diciembre, Febrero y Marzo) tienden a tener menos ataques.

In [1472]:
tabla = pivot_table_country_activity_month = df.pivot_table(
    index=['country', 'activity'],
    columns='month',
    values='year',
    aggfunc='count',
    fill_value=0
)
tabla['most_dangerous_month'] = tabla.idxmax(axis=1)
tabla['incident_count_in_that_month'] = tabla.max(axis=1)
resultado = tabla[['most_dangerous_month', 'incident_count_in_that_month']]
resultado = resultado.sort_values(by='incident_count_in_that_month', ascending=False)
print(resultado)

month                 most_dangerous_month  incident_count_in_that_month
country     activity                                                    
mexico      swimming                     7                            16
spain       swimming                     7                            14
peru        swimming                    10                            10
japan       surf                         9                             9
cuba        wreck                        9                             9
thailand    fishing                      8                             8
croatia     fishing                      7                             7
spain       diving                       7                             7
japan       fishing                      7                             7
            other                        7                             7
            swimming                     7                             7
croatia     other                        7         

Nos queda claro que julio es de los más peligrosos en casi todos los paises en los que trabajamos.

→Reforzar medidas de seguridad en el mes de Julio, ya que es cuando se registran más incidentes, especialmente en actividades como Swimming y Fishing.



In [1473]:
df.to_excel('AquaVentura.xlsx', index=False)